### ISSUE TITLE
Bug in handling of super()

### ISSUE THREAD

**Issue by Retsam** [babel/babel#2048]
The following code exhibits a bug in Babel's handling of `super`:

```
function A () {
  console.log("A");
}

function B () {
    A.call(this);
    console.log("B");
}

B.prototype = Object.create(A.prototype);
B.constructor = B;

class C extends B {
    constructor(data) {
        super({}); //Should call B(), no?
        console.log("C");
    }
}

new C();
```

I'd expect the output here to be "A", "B", and "C", and this is the output given by [ES6Fiddle](http://www.es6fiddle.net/ice18fuo/), built on traceur.  However, with Babel, the output is "A", "C".  The `super()` call in C is calling A's constructor instead of B's.

**Comment by loganfsmyth** [comment]
Sorry, 

    B.constructor = B;

should be

    B.prototype.constructor = B;

or even better

```
B.prototype = Object.create(A.prototype, {
    constructor: {
        value: B,
        enumerable: false,
        writable: true,
        configurable: true
    }
});
```

**Comment by Retsam** [comment]
While I'm sure your alternatives would work; I don't believe you're handling the existing code according to spec.  

Apologies if I'm misreading this, I don't try to read the spec very often, but according to [12.3.5.2](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-getsuperconstructor) it looks like the effect of evaluating `super()` should be to equivalent to evaluating the result of `activeFunction.[[GetPrototypeOf]]()`.  If I understand this correctly, `activeFunction.[[GetPrototypeOf]]` is equivalent to `Object.getPrototypeOf(activeFunction)`, and `activeFunction` here is `C`, and `Object.getPrototypeOf(C) === B`, the spec seems to support the conclusion that evaluating `super()` within the constructor of `C` should call `B`.

**Comment by sebmck** [comment]
You need to a set the `prototype.constructor` property of `B` to `B`, see [node](https://github.com/nodejs/node/blob/master/lib/util.js#L736-743) for the most common/proper way to inherit from a super class.

**Comment by Retsam** [comment]
I'm not disagreeing that B is inheriting from A incorrectly.  But I still believe that, according to the spec, C calling `super` should still be calling B, not A.

**Comment by sebmck** [comment]
The prototype of `C` is `A` because of the incorrect inheritance though.

**Comment by fkling** [comment]
@Retsam: What everybody is trying to say is that you have a mistake in your code. You are doing the inheritance incorrectly. If you do 

```js
B.constructor = B;
```

then `B.prototype.constructor` will *still* point to `C`, so `super` will call `C`. However, if you *correctly* change `B.prototype.constructor` to

```js
B.prototype.constructor = B;
```

then it [works as expected](https://babeljs.io/repl/#?experimental=false&evaluate=true&loose=false&spec=false&code=function%20A%20()%20%7B%0A%20%20console.log(%22A%22)%3B%0A%7D%0A%0Afunction%20B%20()%20%7B%0A%20%20%20%20A.call(this)%3B%0A%20%20%20%20console.log(%22B%22)%3B%0A%7D%0A%0AB.prototype%20%3D%20Object.create(A.prototype)%3B%0AB.prototype.constructor%20%3D%20B%3B%0A%0Aclass%20C%20extends%20B%20%7B%0A%20%20%20%20constructor(data)%20%7B%0A%20%20%20%20%20%20%20%20super(%7B%7D)%3B%20%2F%2FShould%20call%20B()%2C%20no%3F%0A%20%20%20%20%20%20%20%20console.log(%22C%22)%3B%0A%20%20%20%20%7D%0A%7D%0A%0Anew%20C()%3B&playground=false).

**Comment by fkling** [comment]
@Retsam: Even if that was the case (too lazy to check the spec right now), some transpiled things can only work if everybody plays by the rules.

User land code does not have the luxury to access the internal properties of function objects.

**Comment by Retsam** [comment]
@sebmck Actually, no, `Object.getPrototypeOf(C) === B`, which is why I definitely believe that Babel *could* follow the spec here, it were changed.

I'm just failing to understand why "your code has a bug" (a true statement, I'm not disagreeing) is an acceptable reason to not follow the ES6 Spec.  It's clearly not impossible to do so, the traceur transpiler is managing to do it.  (See my original ES6Fiddle link)

**Comment by loganfsmyth** [comment]
Just to clarify, you are asking why

    _get(Object.getPrototypeOf(C.prototype), "constructor", this).call(this, {})

is not instead

    Object.getPrototypeOf(C).call(this, {})

which would work for your code.

One downside is that this would rely on `setPrototypeOf` or `__proto__` assignment support, which is unavailable in older browsers. By using the standard ES5 extension method with class syntax, people are able to use ES6 classes with existing ES5 classes, and they will work on older browsers.

**Comment by sebmck** [comment]
Reopening until someone has verified this on native implementations and read through the spec.

**Comment by lukescott** [comment]
Having to support IE9+, I would want to keep that working. At the same time, the behavior should be consistent across the board. 

Also, as others have pointed out, this seems broken:

```
B.prototype = Object.create(A.prototype);
B.constructor = B;
```

It does work in loose mode though. Which makes me wonder if this would work:

```
(Object.getPrototypeOf(C) || _B).call(this, {})
```

Although I have no idea what is or isn't spec compliant.

**Comment by sebmck** [comment]
Fixed in `development` branch due out for 6.0.
